src/dram_controller.cc

```diff
@@ -552,21 +552,46 @@ bool MEMORY_CONTROLLER::add_wq(const request_type& packet)
   return false;
 }
 
+unsigned long DRAM_ADDRESS_MAPPING::swizzle_bits(champsim::address address, unsigned long segment_size, unsigned long segment_offset, unsigned long field, unsigned long field_bits) const {
+  champsim::address row = champsim::address{get_row(address)};
+  unsigned long permute_field = field;
+
+  unsigned long r_bits = champsim::size(get<SLICER_ROW_IDX>(address_slicer));
+  for(unsigned long r = 0; r < r_bits; r += segment_size) {
+    permute_field ^= row.slice(champsim::dynamic_extent{champsim::data::bits{segment_offset + field_bits},champsim::data::bits{segment_offset}}).to<unsigned long>();
+    row = champsim::address{row.to<unsigned long>() >> segment_size};
+  }
```

Comment:
```suggestion
  for(champsim::dynamic_extent subextent{champsim::data::bits{segment_offset}, field_bits}; subextent.upper <= row.upper_extent(); subextent = champsim::dynamic_extent{subextent.upper, field_bits}) {
    permute_field ^= row.slice(subextent).to<unsigned long>();
  }
```
Again, you've taken a detour through the integral types when what you want to do can be achieved in the high-level interface. Also, I've here used an alternate constructor for `champsim::dynamic_extent` that takes a lower bit and a size.